plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/KotlinMetadataDebuggerCacheService.kt

```diff
@@ -0,0 +1,168 @@
+// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+
+package org.jetbrains.kotlin.idea.debugger.core
+
+import com.google.gson.Gson
+import com.google.gson.JsonSyntaxException
+import com.intellij.debugger.engine.DebugProcess
+import com.intellij.debugger.engine.evaluation.EvaluationContext
+import com.intellij.debugger.engine.evaluation.EvaluationContextImpl
+import com.intellij.debugger.impl.DebuggerManagerListener
+import com.intellij.debugger.impl.DebuggerSession
+import com.intellij.debugger.impl.DebuggerUtilsImpl
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.rt.debugger.MetadataDebugHelper
+import com.sun.jdi.ReferenceType
+import com.sun.jdi.StringReference
+import com.sun.jdi.Value
+import org.jetbrains.annotations.ApiStatus
+import org.jetbrains.kotlin.idea.debugger.base.util.wrapEvaluateException
+import org.jetbrains.kotlin.idea.debugger.base.util.wrapIllegalArgumentException
+import kotlin.io.encoding.Base64
+import kotlin.io.encoding.ExperimentalEncodingApi
+import kotlin.metadata.jvm.KotlinClassMetadata
+
+@ApiStatus.Internal
+@Service(Service.Level.PROJECT)
+class KotlinMetadataDebuggerCacheService private constructor(project: Project) {
```

Comment:
Is it an ok implementation of a cache? Or should I use like something like this instead:
```
  val cache = vmProxy.getOrCreateUserData(METADATA_CACHE_KEY) {
    KotlinMetadataCache(debugProcess)
  }
```

Reply:
Caching in `vmProxy` is better